---
title: "Making Biological Music with Bioconductor"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# install.packages("tidyverse")
library(tidyverse)

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")

# BiocManager::install("SummarizedExperiment")
# BiocManager::install("airway")
# BiocManager::install("ALL")
# BiocManager::install("limma")
# BiocManager::install("biomaRt")

library(SummarizedExperiment)
library(airway)
library(ALL)
library(limma)
library(biomaRt)
```

## Guides

“Orchestrating high-throughput genomic analysis with Bioconductor” - https://www.nature.com/articles/nmeth.3252

"SummarizedExperiment for Coordinating Experimental Assays, Samples, and Regions of Interest" - 
https://bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html

## SummarizedExperiment

```{r}
data(airway, package="airway")
?airway

se <- airway
se

# Accessing

# obtain count data

?assay

assay(se, "counts")
# or
assays(se)$counts

# how many rows and columns do we have?
dim(se)

# View information related to genes (if any)
rowData(se)

# View information related to samples
colData(se)

# how many treatment groups do we have
colData(se)$dex
se$dex

# View metadata
metadata(se)

# Subsetting

# only view first 5 samples
se[ , 1:5] %>% assay
assay(se[ , 1:5])

# only view first 10 genes 
se[1:10, ]

# view first 5 samples and first 10 genes
se[1:10, 1:5]

# save subset
se_subset <- se[1:10, 1:5]

# view count data
assay(se_subset)

# view sample information
colData(se_subset)

# subset for only those samples treated with dexamethasone
se[ , se$dex == "trt"] %>% colData

# notice how the various parts of this object are connected

# modify se object
# add today's date as a new column
colData(se)$date <- "10/18/2023"

# take a look at se
se


# take a look at sample information
colData(se)
# remove this new column
colData(se)$date <- NULL
colData(se)

# compute mean gene counts and add this information to se
rowData(se)$mean_counts <- rowMeans(assay(se, "counts"))
rowData(se)

# use this information to filter out genes that aren't detected in any sample
se[rowData(se)$mean_counts > 0, ]

# log transform the data and create a new assay
assays(se)$log_counts <- log2(assay(se, "counts"))

# show this assay
se
assay(se, "log_counts")
```

## Differential expression analysis example

```{r}
# load ALL data
?ALL
data(ALL)

ALL

ALL_se <- makeSummarizedExperimentFromExpressionSet(ALL)
ALL_se

# Take a look at sample information
colData(ALL_se)
# Only keep B and T cells (refer to column "BT")
ALL_bt <- ALL_se[ , ALL_se$BT %in% c("B", "T")]
colData(ALL_bt)

# Perform differential expression analysis comparing B and T groups
ALL_lm <- assay(ALL_bt, 1)
group <- factor(ALL_bt$BT,
                levels = c("B", "T"))
design <- model.matrix(~group)
colnames(design) <- c("mean_B", "T_vs_B")

fit <- lmFit(ALL_lm, design)
efit <- eBayes(fit)

ALL_lm_results <- topTable(efit, number = 5, coef = "T_vs_B")


# get gene symbols
mart <- useMart("ENSEMBL_MART_ENSEMBL")
mart <- useDataset("hsapiens_gene_ensembl" , mart)
affy_to_gene <- getBM(
  mart = mart,
  filter = "affy_hg_u95av2",
  attributes = c("affy_hg_u95av2", "external_gene_name"),
  values = rownames(ALL_lm_results)
)

ALL_lm_results %>% 
  tibble::rownames_to_column("affy_hg_u95av2") %>% 
  dplyr::inner_join(affy_to_gene, by = "affy_hg_u95av2")

```

## Next steps

QFeatures for quantitative mass spectrometry data - 
https://bioconductor.org/packages/release/bioc/vignettes/QFeatures/inst/doc/Visualization.html

"Exploring QFeatures" (video) - https://www.youtube.com/live/i8gMocvS1KE?si=QRQDvMSLjssen5XR

DESeqDataSet for RNAseq analysis - 
https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

"How to Perform RNA-seq Analysis" (video) - https://www.youtube.com/live/ZUnX_jmffRU?si=FnD-1v2gTPFL_xX6

MultiAssayExperiment for multi-omics experiments -
https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html#creating-a-multiassayexperiment-object-a-rich-example




